ocaml-lsp-server/src/fmt.ml

```diff
@@ -126,7 +126,182 @@ let exec state bin args stdin =
   | Unix.WEXITED 0 -> Result.Ok res.stdout
   | _ -> Result.Error (Unexpected_result { message = res.stderr })
 
-let run state doc : (string, error) result Fiber.t =
+let insert ~line_sep ~line added_lines =
+  let start = { Position.character = 0; line } in
+  let end_ = { Position.character = 0; line } in
+  let range = { Range.start; end_ } in
+  let newText =
+    (added_lines |> Array.to_list |> String.concat ~sep:line_sep) ^ line_sep
+  in
+  { TextEdit.newText; range }
+
+let replace ~line_sep ~line deleted_lines added_lines =
+  let start = { Position.character = 0; line } in
+  let end_ =
+    { Position.character = 0; line = line + Array.length deleted_lines }
+  in
+  let range = { Range.start; end_ } in
+  let newText =
+    (added_lines |> Array.to_list |> String.concat ~sep:line_sep) ^ line_sep
+  in
+  { TextEdit.newText; range }
+
+let delete ~line deleted_lines =
+  let start = { Position.character = 0; line } in
+  let end_ =
+    { Position.character = 0; line = line + Array.length deleted_lines }
+  in
+  let range = { Range.start; end_ } in
+
+  let newText = "" in
+  { TextEdit.newText; range }
+
+module Simple_diff = struct
+  (* based on *)
+  (* https://github.com/paulgb/simplediff/blob/031dc772ca6795cfdfed27384a6b79e772213233/python/simplediff/__init__.py *)
+
+  type item = string
+
+  type diff =
+    | Deleted of item array
+    | Added of item array
+    | Equal of item array
+
+  module StringMap = Map.Make (Stdlib.String)
+  module IntMap = Map.Make (Stdlib.Int)
```

Comment:
`String.Map` and `Int.Map`. This might require editing `Import` to add the necessary imports.